src/ValidationContext.php

```diff
@@ -6,42 +6,66 @@
 
 use RuntimeException;
 use Yiisoft\Arrays\ArrayHelper;
+use Yiisoft\Validator\Rule\Each;
+use Yiisoft\Validator\Rule\Nested;
 
 /**
  * Validation context that might be taken into account when performing validation.
+ * The context is passed to both direct calls of {@see ValidatorInterface::validate()} and when {@see Nested}
+ * or {@see Each} is used.
  *
  * @psalm-import-type RulesType from ValidatorInterface
  */
 final class ValidationContext
 {
+    /**
+     * @var ValidatorInterface|null A validator instance. Null means context data was not set yet
+     * with {@see setContextDataOnce()}.
+     */
     private ?ValidatorInterface $validator = null;
 
     /**
-     * @var mixed The raw validated data.
+     * @var mixed The raw validated data. Null means context data was not set yet with {@see setContextDataOnce()}.
      */
     private mixed $rawData = null;
 
     /**
-     * @var DataSetInterface|null Data set the attribute belongs to. Null if data set not set.
+     * @var DataSetInterface|null Data set the attribute belongs to.
+     * Null if data set was not set yet with {@see setDataSet()}.
      */
     private ?DataSetInterface $dataSet = null;
 
     /**
-     * @var string|null Validated attribute name. Null if a single value is validated.
+     * @var string|null Validated data set attribute name. Null if a single value is validated.
      */
     private ?string $attribute = null;
 
+    /**
+     * @var AttributeTranslatorInterface|null Default attribute translator to use if attribute translator is not set.
+     */
     private ?AttributeTranslatorInterface $defaultAttributeTranslator = null;
 
     /**
      * @param array $parameters Arbitrary parameters.
+     * @param AttributeTranslatorInterface|null $attributeTranslator Optional attribute translator instance to use.
+     * If null is provided, or it's not specified, default passed through {@see setContextDataOnce()} is used.
      */
     public function __construct(
         private array $parameters = [],
         private ?AttributeTranslatorInterface $attributeTranslator = null,
     ) {
     }
 
+    /**
+     * Set context data if it is not set yet.
```

Comment:
Add recommendation don't use this method in userland. The following methods needs in other validator implementation or may be in validator decorator:
- `setContextDataOnce()`,
- `setDataSet()`,
- `setAttribute()`.

Reply:
I've marked these as internal methods.